contracts/src/main/kotlin/com/contractsAndStates/contracts/BuildPhaseContract.kt

```diff
@@ -28,273 +27,276 @@ class BuildPhaseContract : Contract {
-                "There should be one output state of type SettlementState" using ((tx.outputsOfType<SettlementState>().size == 1))
-                "There should be one output state of type RoadState" using ((tx.outputsOfType<RoadState>().size == 1))
+                "There should be one output state of type SettlementState" using (outputSettlements.size == 1)
+                "There should be one output state of type RoadState" using (outputRoads.size == 1)
+                "There should be no resource input states" using (inputResources.isEmpty())
+                "There should be one ref input state of type TurnTrackerState" using (turnTrackers.size == 1)
 
                 /**
                  *  ******** BUSINESS LOGIC ********
-                 *  Check that the counter party is proposing a move that is allowed by the rules of the game.
+                 *  Check that the party is proposing a move that is allowed by the rules of the game.
                  */
-
-                val newSettlement = tx.outputsOfType<SettlementState>().single()
-                val hexTileCoordinate = newSettlement.hexTileCoordinate
-                val hexTileIndex = newSettlement.hexTileIndex
-                val turnTracker = tx.referenceInputsOfType<TurnTrackerState>().single()
+                val newSettlement = outputSettlements.single()
+                val onCorner = newSettlement.getAbsoluteCorner()
+                val newRoad = outputRoads.single()
+                val onSide = newRoad.getAbsoluteSide()
+                val turnTracker = turnTrackers.single()
+                val currentPlayer = inputBoard.players[turnTracker.currTurnIndex]
 
                 /**
-                 * Check Settlements - In order to determine whether or not a settlement has been previously placed in a given location,
-                 * we first have to determine which hexTiles lay adjacent to the one in question.
+                 * Check Settlements - no previous settlement should be on the place, on the neighboring corners,
+                 * and for the overlapping ones.
                  */
-
-                // Here we Initialize storage for a relevant list of relevant HexTiles.
-                val relevantHexTileNeighbours: ArrayList<HexTile?> = arrayListOf()
-
-                // Add neighbouring hexTiles to the storage array if they exist.
-                inputGameBoardState.hexTiles.get(hexTileIndex).sides.getNeighborsOn(hexTileCoordinate.getAdjacentSides())
-                        .forEach {
-                            if (it != null) relevantHexTileNeighbours.add(inputGameBoardState.hexTiles.get(it))
+                val itAndNeighboringCorners = inputBoard.hexTiles
+                        .getItAndNeighboringCorners(onCorner)
+                "A settlement must not have previously been built in this vicinity." using
+                        itAndNeighboringCorners.none { inputBoard.settlementsPlaced.hasOn(it) }
```

Comment:
@rogersanick Example of improved readability of the whole lot. Not finished.
@abellagonzalo Example of usage of the traversing functions.